train.py

```diff
@@ -134,41 +134,46 @@ def main(job_config: JobConfig):
         f"{color.red}size: {model_param_count:,} total parameters{color.reset}"
     )
 
-    if parallel_dims.pp_enabled:
-        stages, model_parts = models_pipelining_fns[model_name](
-            whole_model, pp_mesh, parallel_dims, job_config, device, model_config
-        )
-    else:
-        # In 1D/2D cases or PP with simple schedules, model_parts is just one item
-        # for PP with looped schedules, each item is one stage-model-chunk
-        # we iterate all model_parts for applying SPMD parallelism, compilation, optimizer, and checkpointing
-        model_parts = [whole_model]
-
-    # apply PT-D DP/TP parallelisms and activation checkpointing
-    model_parts = [
-        models_parallelize_fns[model_name](m, world_mesh, parallel_dims, job_config)
-        for m in model_parts
-    ]
-
-    init_device = "cpu" if job_config.checkpoint.create_seed_checkpoint else "cuda"
-    for model in model_parts:
-        model.to_empty(device=init_device)
-
-    # loss fn can be shared by pipeline-parallel or non-pp execution
+    # loss function to be shared by Pipeline Parallel and spmd training
     def loss_fn(pred, labels):
         return torch.nn.functional.cross_entropy(
             pred.flatten(0, 1), labels.flatten(0, 1)
         )
 
+    # apply parallelisms and initialization
     if parallel_dims.pp_enabled:
+        # apply PT-D Pipeline Parallel
+        stages, model_parts = models_pipelining_fns[model_name](
+            model, pp_mesh, parallel_dims, job_config, device, model_config
+        )
         pp_schedule = build_pipeline_schedule(
             job_config, parallel_dims, stages, loss_fn
         )
+
+        # For PP with looped schedules, each item in model_parts is one stage-model-chunk.
```

Comment:
i guess one thing that could clean up train.py a bit is to move all this code inside of `models_pipelining_fns` and have it return (pp_schedule, stages, model_parts)

(not sure if we need stages, model_parts in train.py anymore actually).

if we did that, i'd update the comment "apply PT-D Pipeline Parallel" to mention also applying PP/DP/TP

Reply:
we don't need `stages` any more after `build_pipeline_schedule`, but we still need `model_parts` (for optimizer, checkpointing, fp8 updates, etc.)

I think it's clearer we separate PP and other `parallelize_llama`

Reply:
Yeah I think it makes sense to merge `build_pipeline_schedule` into `pipeline_llama`. Do you want me to do that in this PR?